api/v1alpha3/awsmachine_webhook.go

```diff
@@ -29,3 +35,50 @@ func (r *AWSMachine) SetupWebhookWithManager(mgr ctrl.Manager) error {
 		For(r).
 		Complete()
 }
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha3-awsmachine,mutating=false,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=awsmachines,versions=v1alpha3,name=validation.awsmachine.infrastructure.cluster.x-k8s.io
+
+var _ webhook.Validator = &AWSMachine{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachine) ValidateCreate() error {
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachine) ValidateUpdate(old runtime.Object) error {
+	newAWSMachine, err := runtime.DefaultUnstructuredConverter.ToUnstructured(r)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert new AWSMachine to unstructured object")
+	}
+	oldAWSMachine, err := runtime.DefaultUnstructuredConverter.ToUnstructured(old)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert old AWSMachine to unstructured object")
+	}
+
+	newAWSMachineSpec := newAWSMachine["spec"].(map[string]interface{})
+	oldAWSMachineSpec := oldAWSMachine["spec"].(map[string]interface{})
+
+	// allow changes to providerID
+	delete(oldAWSMachineSpec, "providerID")
+	delete(newAWSMachineSpec, "providerID")
+
+	// allow changes to additionalTags
+	delete(oldAWSMachineSpec, "additionalTags")
+	delete(newAWSMachineSpec, "additionalTags")
+
+	// allow changes to additionalSecurityGroups
+	delete(oldAWSMachineSpec, "additionalSecurityGroups")
+	delete(newAWSMachineSpec, "additionalSecurityGroups")
+
+	if !reflect.DeepEqual(oldAWSMachineSpec, newAWSMachineSpec) {
+		return field.Forbidden(field.NewPath("spec"), "cannot be modified")
```

Comment:
I just learned about a better way to return validation errors - please see https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html / xref https://github.com/kubernetes-sigs/cluster-api/issues/1795 too.

Reply:
@ncdc sorry for being late. I updated the pr, ptal :)